### ISSUE TITLE
Enable use of 'react-intl' without React Context

### ISSUE THREAD

**Issue by alexharri** [formatjs/formatjs#4192]
## Use case

Currently, `react-intl` components such as `FormattedMessage` cannot be used in React Server Components because of their use of React Context.

The easiest way to encounter this is in a Next.js app using the new App Router.

```tsx
import { FormattedMessage } from "react-intl";

function ServerComponent() {
  return <FormattedMessage id="example" />;
}
```

Trying to use this Server Component we encounter the following exception:

```
Error: createContext only works in Client Components.

Add the "use client" directive at the top of the file to use it. Read more: https://nextjs.org/docs/messages/context-in-server-component
```

## Potential solution

Currently, components like `FormattedMessage` expect to receive an `intl` instance from `useIntl`:

```tsx
function FormattedMessage(props) {
  const intl = useIntl();
  // ...
}
```

The use of React Contexts can be avoided by introducing "contextless alternatives" that receive an `intl` instance as a prop:

```tsx
function FormattedMessageImplementation(props) {
  const intl = props.intl;
  // ...
}

function FormattedMessage(props) {
  const intl = useIntl();
  return <FormattedMessageImplementation {...props} intl={intl} />;
}
```

The contextless alternative could be exposed through a submodule of `react-intl`:

```tsx
import { FormattedMessage } from "react-intl/without-context";
```

This is similar to how `react-i18next` exposes contextless the `Trans`:

```tsx
import { Trans } from "react-i18next/TransWithoutContext";
```

The submodule would also need to expose functions such as `createIntl` because the use of `React.createContext` in the `react-intl/index` barrel causes the same `createContext only works in Client Components` error to occur when importing `createIntl` from `react-intl`.

```tsx
// This encounters the 'createContext only works in Client Components' error

import { createIntl } from "react-intl";

function ServerComponent() {
  const intl = createIntl(...);
}
```

## Alternatives considered

I tried using `createIntl` from `@formatjs/intl` and using that in Server Components:

```tsx
import { createIntl } from "@formatjs/intl";

function ServerComponent() {
  const intl = createIntl({ locale, messages });

  return <p>{intl.formatMessage({ id: "example" })}</p>;
}
```

This approach works for simple messages. However, using `values` encounters type and runtime errors:

```tsx
import { createIntl } from "@formatjs/intl";

function ServerComponent() {
  const intl = createIntl({ locale, messages });

  // Type error: Type 'Element' is not assignable to type 'string | string[]'.
  return (
    <p>
      {intl.formatMessage(
        { id: "example" },
        {
          p: (chunks) => <p>{chunks}</p>,
          code: (chunks) => <code>chunks</code>,
        },
      )}
    </p>
  );
}
```

During runtime, the following warnings are encountered:

```
Warning: Each child in a list should have a unique "key" prop.

Check the top-level render call using <p>. See https://reactjs.org/link/warning-keys for more information.
    at code

Warning: Each child in a list should have a unique "key" prop. See https://reactjs.org/link/warning-keys for more information.
    at p
```

The `createIntl` function from `react-intl` handles these key problems.

## Conclusion

It's currently not possible to use `react-intl` in Server Components.

I would love to hear your thoughts on the proposed solution. I would also be more than willing to help implement this and create a PR.

**Comment by longlho** [comment]
I think using the imperative API is the right solution. You can use `createIntl` from `react-intl` to fix that issue

**Comment by AugustinMauroy** [comment]
Also to accomplish this issue we need to specify that `<FormmatedMessage />` is client component

**Comment by alexharri** [comment]
> I think using the imperative API is the right solution. You can use `createIntl` from `react-intl` to fix that issue

Using `createIntl` would be a viable solution, though it's currently not possible in Server Components:

```tsx
import { createIntl } from "react-intl";

export function ServerComponent() {
  const intl = createIntl({ ... });
  // ...
}

// When run, the following error is encountered:
//
// Error: createContext only works in Client Components.
// Add the "use client" directive at the top of the file to use it...
```

Adding `"use client";` to `react-intl/index.js` changes the error encountered to the following:

```
Error: Attempted to call createIntl() from the server but
createIntl is on the client. It's not possible to invoke a
client function from the server, it can only be rendered as
a Component or passed to props of a Client Component.
``` 

I also tried attempting to directly import `src/components/provider` (where `createIntl` is defined):

```tsx
import { createIntl } from "react-intl/src/components/provider";

export function ServerComponent() {
  const intl = createIntl({ ... });
  // ...
}
```

But it also encounters the `createContext` error (because of the `injectIntl` import):

```
Error: createContext only works in Client Components.
Add the "use client" directive at the top of the file to use it...
```

This happens because `IntlProvider` is defined in the same module, which depends on `injectIntl`, which uses React context.

---

We could resolve this by moving `createIntl` into a separate module containing no use of context, and exposing it via a submodule of `react-intl`

```tsx
// Could be called something else
import { createIntl } from "react-intl/server";
```

The main drawback to not supporting `FormattedMessage` (and related components) in Server Components as well is that when users convert existing components to Server Components, they would need to migrate usage of `FormattedMessage` to the imperative API.

**Comment by am1rb** [comment]
any updates on this issue?

**Comment by longlho** [comment]
Is there any repro codesandbox to replicate this issue?

**Comment by am1rb** [comment]
hey @longlho, here is an example: https://codesandbox.io/p/devbox/ecstatic-cloud-2rr4xd?file=%2Fapp%2Fpage.tsx%3A5%2C80

It's enough to try to import `createIntl` in a react server component; you will face an error in your dev machine.

**Comment by slaivyn** [comment]
Hello @longlho @am1rb ,
That's strange but in version6.6.1 it still does not work for me...

@am1rb I presume that your fix works on your side but I don't understand why I still have the React Context Error (`Error: createContext only works in Client Components`) as my use case is very simple.

The line causing the error is:
```typescript
<FormattedMessage id="common.submit" />
``` 

With FormattedMessage imported like this:
```typescript
import { FormattedMessage } from 'react-intl';
``` 

In a simple NextJS page (server component

**Comment by slaivyn** [comment]
Okay I have found out the reason...
Here is a repro:

```typescript
import { FormattedDate } from "react-intl";
import ServerIntlProvider from "./server-intl-provider";

const Page = async () => {
  return (
    <ServerIntlProvider messages={{}} locale={"fr"}>
      <FormattedDate value={lastSession.openedAt} />
    </ServerIntlProvider>
  );
};
export default Page;
``` 

With ServerIntlProvider.tsx:
```typescript
'use client';

import { IntlProvider } from 'react-intl';

export default function ServerIntlProvider({ messages, locale, children }) {
  return (
    <IntlProvider messages={messages} locale={locale}>
      {children}
    </IntlProvider>
  );
}
``` 

To have it work I have to import the react-intl components from a "client proxy" file like this one:
```typescript
'use client';
export * from 'react-intl';
```

**Comment by atarek12** [comment]
what are the drawbacks of using the `<FormattedMessage />` as a client component? 
I like the API of the  `<FormattedMessage />` but on the other side, we could use `createIntl` on the server component and use `intl.formatMessage()` instead. 
which is better?

**Comment by am1rb** [comment]
Hey @slaivyn, `<FormattedMessage  />` uses React Context under the hood and with the current implementation, there is no way to use the component and also `useIntl`  in RSC.

I recently found [this section](https://github.com/reactjs/rfcs/blob/main/text/0188-server-components.md#sharing-code-between-server-and-client) on RSC rfs and think by using [Conditional exports](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#react-server-conditional-exports) we can make react-intl more easy to use.

I see next-intl took the same approach for providing the same APIs (for example`useTranslations`) on both client and server components: https://next-intl-docs.vercel.app/docs/environments/server-client-components

**Comment by am1rb** [comment]
> Okay I have found out the reason... Here is a repro:
> 
> ```ts
> import { FormattedDate } from "react-intl";
> import ServerIntlProvider from "./server-intl-provider";
> 
> const Page = async () => {
>   return (
>     <ServerIntlProvider messages={{}} locale={"fr"}>
>       <FormattedDate value={lastSession.openedAt} />
>     </ServerIntlProvider>
>   );
> };
> export default Page;
> ```
> 
> With ServerIntlProvider.tsx:
> 
> ```ts
> 'use client';
> 
> import { IntlProvider } from 'react-intl';
> 
> export default function ServerIntlProvider({ messages, locale, children }) {
>   return (
>     <IntlProvider messages={messages} locale={locale}>
>       {children}
>     </IntlProvider>
>   );
> }
> ```
> 
> To have it work I have to import the react-intl components from a "client proxy" file like this one:
> 
> ```ts
> 'use client';
> export * from 'react-intl';
> ```

@slaivyn the solution you shared does not work properly because it marks the component and all of its children as client components

**Comment by am1rb** [comment]
> what are the drawbacks of using the `<FormattedMessage />` as a client component? I like the API of the `<FormattedMessage />` but on the other side, we could use `createIntl` on the server component and use `intl.formatMessage()` instead. which is better?

I would like to implement my components in a way that I could reuse them on both client and server, so I prefer not to tighten them with client APIs. but at the moment, I see no way cause the `useIntl` hook also uses Context API under the hood, and there is a similar problem with it.

I guess we need a new issue to extend the lib if possible and make the `useIntl` and maybe `<FormattedMessage />` compatible with RSC environment. I'll check the issues list today and create one if we don't have it.

cc @longlho

**Comment by joshunger** [comment]
@am1rb is there a new issue we should subscribe to?

**Comment by adelin-b** [comment]
Hi, the current state for nextjs and react-intl is now not very good, do you think we can just use a "use client" wrapper around FormattedMessage and then a fix will come so we have minimal changes ? The default message and extraction system of react-intl is really a good DX but it don't seem to work as well as next-intl atm.

**Comment by doksara** [comment]
@am1rb I'm still getting the `Error: createContext only works in Client Components `even after #4316 , how is that possible?

**Comment by carlesba** [comment]
Did we figure this out? Is there any official way to use react-intl with React Server Components?

**Comment by krusk8** [comment]
ANy Updates on it? I think it should will be enable the possibility to use the translated Content also to "generateMeta" in NextJS App routing and also on Middleware or Server side components. Someone have found a solution for that? Or some other library that work better than ract-intl?

**Comment by krusk8** [comment]
The solution is to update version "react-intl" to "6.6.1" and import directly from file
`import { createIntl } from "react-intl/src/components/createIntl"`

thanks to PR of @am1rb  https://github.com/formatjs/formatjs/pull/4316

**Comment by longlho** [comment]
Use @formatjs/intl if you want ssr
